presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java

```diff
@@ -15,283 +15,367 @@
+            List<RemoteSourceNode> remoteSources = fragment.getRemoteSourceNodes();
+            checkArgument(remoteSources.size() == 2, "two remote sources are expected, got: %s", remoteSources.size());
+            RemoteSourceNode firstRemoteSource = remoteSources.get(0);
+            RemoteSourceNode secondRemoteSource = remoteSources.get(1);
+            JavaPairRDD<Integer, PrestoSparkRow> firstRdd = rddInputs.get(firstRemoteSource.getSourceFragmentIds().get(0));
+            JavaPairRDD<Integer, PrestoSparkRow> secondRdd = rddInputs.get(secondRemoteSource.getSourceFragmentIds().get(0));
+            FlatMapFunction2<Iterator<Tuple2<Integer, PrestoSparkRow>>, Iterator<Tuple2<Integer, PrestoSparkRow>>, Tuple2<Integer, PrestoSparkRow>> taskProcessor =
+                    createTaskProcessor(
+                            executorFactoryProvider,
+                            serializedTaskDescriptor,
+                            firstRemoteSource.getId().toString(),
+                            secondRemoteSource.getId().toString(),
+                            taskStatsCollector);
+            return JavaPairRDD.fromJavaRDD(
+                    firstRdd.zipPartitions(
+                            secondRdd,
+                            taskProcessor));
         }
 
-        private static List<List<ScheduledSplit>> assignSplitsToTasks(List<ScheduledSplit> scheduledSplits, int numTasks)
-        {
-            List<List<ScheduledSplit>> assignedSplits = new ArrayList<>();
-            for (int i = 0; i < numTasks; i++) {
-                assignedSplits.add(new ArrayList<>());
-            }
+        throw new IllegalArgumentException(format("unsupported number of inputs: %s", rddInputs.size()));
+    }
 
-            for (ScheduledSplit split : scheduledSplits) {
-                int taskId = Objects.hash(split.getPlanNodeId(), split.getSequenceId()) % numTasks;
-                if (taskId < 0) {
-                    taskId += numTasks;
-                }
+    private JavaPairRDD<Integer, PrestoSparkRow> createSourceRdd(
+            JavaSparkContext sparkContext,
+            Session session,
+            PlanFragment fragment,
+            PrestoSparkTaskExecutorFactoryProvider executorFactoryProvider,
+            CollectionAccumulator<SerializedTaskStats> taskStatsCollector,
+            TableWriteInfo tableWriteInfo)
+    {
+        // TODO: Possible in case of a broadcast join
+        checkArgument(fragment.getRemoteSourceNodes().isEmpty(), "source task with remote sources is not supported");
+
+        List<TableScanNode> tableScans = findTableScanNodes(fragment.getRoot());
+        checkArgument(
+                tableScans.size() == 1,
+                "exactly one table scan is expected in SOURCE_DISTRIBUTION fragment. fragmentId: %s, actual number of table scans: %s",
+                fragment.getId(),
+                tableScans.size());
+        verify(tableScans.size() == fragment.getTableScanSchedulingOrder().size());
```

Comment:
hmm. maybe we can make the previous `checkArugment` to assert both `tableScan.size() == 1 && fragment.getTableScanSchedulingOrder().size() == 1`?

Reply:
If the number of `tableScanSchedulingOrder` is not equal to the number of table scans that means that the plan generation is somehow broken, thus it feels like an assertion

Reply:
@arhimondr : But the previous assertion already asserts `tableScans.size()` is always 1...

Reply:
Actually yeah, let me simply remove this assertion